docs/jmespath-grammar.txt

```diff
@@ -0,0 +1,107 @@
+dot-separator   = ws "." ws
+pipe-separator  = ws "|" ws
+or-separator    = ws "||" ws
+and-separator   = ws "&&" ws
+name-separator  = ws ":" ws
+value-separator = ws "," ws
+lt              = ws "<" ws
+lte             = ws "<=" ws
+gt              = ws ">" ws
+gte             = ws ">=" ws
+eq              = ws "==" ws
+ne              = ws "!=" ws
+expref-token    = "&" ws
+not-token       = "!" ws
+
+; "&&" binds more tightly than "||"
+; "||" binds more tightly than "|".
+terminal     = pipe / or / and
+non-terminal = root-expr / wildcard-values
+pipe         = expr pipe-separator (non-terminal / or / and)
+or           = (non-terminal / and) or-separator (non-terminal / or / and)
+and          = non-terminal and-separator (non-terminal / and)
+
+; subexpr handles most that descend into nodes.
+subexpr            = object-subexpr / array-subexpr
+object-subexpr     = object-subexpr-lhs dot-separator object-subexpr-rhs
+object-subexpr-lhs = subexpr / index / flatten / wildcard-index /
+                     filter / identifier / current-node / literal /
+                     multi-hash / function / group / slice / wildcard-values
+object-subexpr-rhs = identifier / multi-list / multi-hash / function /
+                     wildcard-values
+array-subexpr      = array-subexpr-lhs array-subexpr-rhs
+array-subexpr-lhs  = subexpr / index / flatten / wildcard-index / filter /
+                     identifier / current-node / literal / root-multi-list /
+                     function / group / slice / wildcard-values / literal
+array-subexpr-rhs  = index / slice / wildcard-index / flatten / filter
+
+; Array related rules
+index           = open-bracket number close-bracket
+slice           = open-bracket *number name-separator
+                               *number *(name-separator *number)
+                               close-bracket
+wildcard-index   = open-bracket "*" close-bracket
+flatten          = open-bracket close-bracket
+filter           = open-bracket "?" filter-condition close-bracket
+filter-condition = not / non-test / terminal / comparison
+
+not             = not-token (non-test / not)
+comparison      = non-terminal comparator (non-test / not / wildcard-values)
+comparator      = lt / lte / gt / gte / eq / ne
+
+root-multi-list = open-bracket
+                  (root-expr / multiple-values / terminal)
+                  close-bracket
+multi-list      = open-bracket (expr / multiple-values) close-bracket
+multiple-values = expr 1*( value-separator expr )
+
+multi-hash      = open-brace (keyval *(value-separator keyval)) close-brace
+keyval          = identifier name-separator expr
+
+function        = unquoted-string arg-list
+arg-list        = open-group *1(arg *(value-separator arg)) close-group
+arg             = expr / expref
+
+expref          = expref-token expr
+wildcard-values = "*"
+current-node    = "@"
+
+number            = *"-" 1*DIGIT
+literal           = "`" 1*(unescaped-literal / escaped-literal) "`"
```

Comment:
I thought `[` was valid (currently) because JMESPath allows you to pass in unquoted strings here. It will try to parse and see that it failed and then parse it as `"["`.